cpp/oneapi/dal/detail/memory_impl_dpc.hpp

```diff
@@ -19,58 +19,48 @@
 #ifdef ONEDAL_DATA_PARALLEL
 #include <cstring>
 
-#include "oneapi/dal/detail/common_dpc.hpp"
 #include "oneapi/dal/detail/policy.hpp"
 
 namespace oneapi::dal::detail {
 namespace v1 {
 
+ONEDAL_EXPORT void* malloc(const data_parallel_policy&, std::size_t size, const sycl::usm::alloc&);
+ONEDAL_EXPORT void free(const data_parallel_policy&, void* pointer);
+ONEDAL_EXPORT void fill(const data_parallel_policy&,
+                        void* dest,
+                        std::size_t dest_size,
+                        const void* pattern,
+                        std::size_t pattern_size);
+
 template <typename T>
 inline T* malloc(const data_parallel_policy& policy,
                  std::int64_t count,
                  const sycl::usm::alloc& alloc) {
-    auto& queue = policy.get_queue();
-    auto device = queue.get_device();
-    auto context = queue.get_context();
-    // TODO: is not safe since sycl::memset accepts count as size_t
-    return sycl::malloc<T>(count, device, context, alloc);
+    ONEDAL_ASSERT_MUL_OVERFLOW(std::size_t, sizeof(T), count);
+    const std::size_t bytes_count = sizeof(T) * count;
+    return static_cast<T*>(malloc(policy, bytes_count, alloc));
 }
 
 template <typename T>
 inline void free(const data_parallel_policy& policy, T* pointer) {
     using mutable_t = std::remove_const_t<T>;
-    sycl::free(const_cast<mutable_t*>(pointer), policy.get_queue());
-}
-
-inline void memset(const data_parallel_policy& policy,
-                   void* dest,
-                   std::int32_t value,
-                   std::int64_t size) {
-    // TODO: is not safe since queue.memset accepts size as size_t
-    auto event = policy.get_queue().memset(dest, value, size);
-    event.wait();
+    free(policy, reinterpret_cast<void*>(const_cast<mutable_t*>(pointer)));
 }
 
-inline void memcpy(const data_parallel_policy& policy,
-                   void* dest,
-                   const void* src,
-                   std::int64_t size) {
-    // TODO: is not safe since queue.memcpy accepts size as size_t
-    auto event = policy.get_queue().memcpy(dest, src, size);
-    event.wait();
-}
+ONEDAL_EXPORT void memset(const data_parallel_policy& policy,
+                          void* dest,
+                          std::int32_t value,
+                          std::int64_t size);
+ONEDAL_EXPORT void memcpy(const data_parallel_policy& policy,
+                          void* dest,
+                          const void* src,
+                          std::int64_t size);
```

Comment:
Don't understand logic of function ordering. I would put non-template function first, and than  templates.

Reply:
Reordered